Source/EventFlow.Sql/Connections/SqlConnection.cs

```diff
@@ -143,11 +144,30 @@ public abstract class SqlConnection<TConfiguration, TRetryStrategy, TConnectionF
                 {
                     using (var sqlConnection = await ConnectionFactory.OpenConnectionAsync(Configuration.ConnectionString, cancellationToken).ConfigureAwait(false))
                     {
-                        return await withConnection(sqlConnection, c).ConfigureAwait(false);
+                        try
+                        {
+                            return await withConnection(sqlConnection, c).ConfigureAwait(false);
+                        }
+                        catch (Exception exception)
+                        {
+                            if (IsOptimisticConcurrencyException(exception))
+                            {
+                                throw new OptimisticConcurrencyException(exception.Message, exception);
+                            }
+                            throw;
+                        }
+                        
                     }
                 },
                 label,
                 cancellationToken);
         }
+
+        /// <summary>
+        /// Override this method to detect and convert specific concurrency exceptions 
+        /// to Eventflow.OptimisticConcurrencyException
+        /// This allows reuse of detection in both Event Persistense & ReadModelStores.
+        /// </summary>
+        protected virtual bool IsOptimisticConcurrencyException(Exception exception) => false;
```

Comment:
It cannot be done in such generic way:
Only particular storage mechanism can detect (and therefore raise) OptimisticConcurrencyException...
I.e. for store events ic can be violate unique constraint, for store read model it can be violate primarry key constraint, and violation of unique key constraint can be related with error in read model. And mask this exception as OptimisticConcurrencyException is the wrong way....

Reply:
>primarry key constraint, and violation of unique key constraint 
And seems it not possible to distinguish between this two in all implementations :-) 
Let's assume it's responsibility of read store and make most specific implementation (it's ok for me). What could be done for MS SQL readstore short of checking same exception?

Reply:
Do you mean some kind of: https://github.com/FortisOnline/EventFlow/blob/develop/Source/EventFlow.EntityFramework/Extensions/ExceptionExtensions.cs 
and:
https://github.com/FortisOnline/EventFlow/blob/develop/Source/EventFlow.EntityFramework/DefaultUniqueConstraintDetectionStrategy.cs

Reply:
Let's forget code structure questions for a moment.
Do you agree that catching SqlException with Numbers 2601 || 2627 in when MsSQL upserting read model is correct way to detect concurrency, and arguing just about code structure, or you have something else in mind?

Reply:
I don't think so that check SqlException is good way to detect optimistic concurrency.

Now it checked by affected rows. I think it is the best possible way to detect concurrency during update. To detect race between two inserts (and only inserts) check sql execption may be suitable....

Reply:
I'm affraid situation when someone adds additional unique constraint to table with ReadModel. And updating this read model (without concurrency) will trigger this unique constraint violation.

In this case we will mask original exception with OptimisticConcurrencyException, we will try to update this read model several times (until retry count exceed)...

Reply:
So, let's go with:
Introduce throwing OptimisticConcurrencyException if unique constraint violation detected on inserting (not updating) sql read store

Reply:
Ok. Please don't forget to add comment that it is not always true. May be right inside OptimisticConcurrencyException message.